src/core_editor/line_buffer.rs

```diff
@@ -578,4 +619,87 @@ mod test {
     }
+
+    #[rstest]
+    #[case("This is a test", (0, 13), "This is a tets", (0, 14))]
+    #[case("This is a test", (0, 14), "This is a test", (0, 14))]
+    // FIXME: swap_graphemes does not work with first index
+    // #[case("This is a test", (0, 0), "hTis is a test", (0, 0))]
+    fn swap_graphemes_work(
+        #[case] input: &str,
+        #[case] in_location: (usize, usize),
+        #[case] output: &str,
+        #[case] out_location: (usize, usize),
+    ) {
+        let mut line_buffer = buffer_with(input);
+        line_buffer.set_insertion_point(in_location.0, in_location.1);
+
+        line_buffer.swap_graphemes();
+
+        let mut expected = buffer_with(output);
+        expected.set_insertion_point(out_location.0, out_location.1);
+
+        assert_eq!(line_buffer, expected);
+    }
+
+    #[rstest]
+    #[case("This is a test", (0, 9), "This is test a", (0, 14))]
+    // FIXME: Shouldn't this be a no-op
+    // #[case("This is a test", (0, 14), "This is a test", (0, 14))]
+    fn swap_words_works(
+        #[case] input: &str,
+        #[case] in_location: (usize, usize),
+        #[case] output: &str,
+        #[case] out_location: (usize, usize),
+    ) {
+        let mut line_buffer = buffer_with(input);
+        line_buffer.set_insertion_point(in_location.0, in_location.1);
+
+        line_buffer.swap_words();
+
+        let mut expected = buffer_with(output);
+        expected.set_insertion_point(out_location.0, out_location.1);
+
+        assert_eq!(line_buffer, expected);
+    }
+
+    // #[rstest]
+    // // FIXME: Panics
+    // // #[case("line 1\nline 2", (1, 0), "line 1\nline 2", (0, 0))]
+    // fn moving_up_works(
+    //     #[case] input: &str,
+    //     #[case] in_location: (usize, usize),
+    //     #[case] output: &str,
+    //     #[case] out_location: (usize, usize),
+    // ) {
+    //     let mut line_buffer = buffer_with(input);
+    //     line_buffer.set_insertion_point(in_location.0, in_location.1);
+    //
+    //     line_buffer.move_line_up();
+    //
+    //     let mut expected = buffer_with(output);
+    //     expected.set_insertion_point(out_location.0, out_location.1);
+    //
+    //     assert_eq!(line_buffer, expected);
+    // }
+    //
+    // #[rstest]
+    // // FIXME: results in insertion point as 0,7
+    // // #[case("line 1\nline 2", (0, 0), "line 1\nline 2", (1, 0))]
+    // fn moving_down_works(
+    //     #[case] input: &str,
+    //     #[case] in_location: (usize, usize),
+    //     #[case] output: &str,
+    //     #[case] out_location: (usize, usize),
+    // ) {
+    //     let mut line_buffer = buffer_with(input);
+    //     line_buffer.set_insertion_point(in_location.0, in_location.1);
+    //
+    //     line_buffer.move_line_down();
+    //
+    //     let mut expected = buffer_with(output);
+    //     expected.set_insertion_point(out_location.0, out_location.1);
+    //
+    //     assert_eq!(line_buffer, expected);
+    // }
```

Comment:
I think we have only done a partial move to only string representation which is causing tests to fail. We should outline what all is the expectation here.

Reply:
Can you say a bit more about what's missing? I'm not up to speed